File: src/swift/4_BehavioralPatterns/11_Template/Exercise.swift
import Foundation
import XCTest

class Creature
{
  public var attack, health: Int

  init(_ attack: Int, _ health: Int)
  {
    self.attack = attack
    self.health = health
  }
}

class CardGame
{
  var creatures: [Creature]

  init(_ creatures: [Creature])
  {
    self.creatures = creatures
  }

  // returns the index of the reature that won the fight
  // returns -1 if there is no clear winner (both alive or both dead)
  func combat(_ creature1: Int, _ creature2: Int) -> Int
  {
    let first = creatures[creature1]
    let second = creatures[creature2]
    hit(first, second)
    hit(second, first)
    let firstAlive = (first.health > 0)
    let secondAlive = (second.health > 0)
    if firstAlive == secondAlive
    {
      return -1
    }
    return firstAlive ? creature1 : creature2
  }

  internal func hit(_ attacker: Creature, _ other: Creature)
  {
    precondition(false, "this method needs to be overridden")
  }
}

class TemporaryCardDamageGame : CardGame
{
  override func hit(_ attacker: Creature, _ other: Creature)
  {
    let oldHealth = other.health
    other.health -= attacker.attack
    if other.health > 0
    {
      other.health = oldHealth
    }
  }
}

class PermanentCardDamage : CardGame
{
  override func hit(_ attacker: Creature, _ other: Creature)
  {
    other.health -= attacker.attack
  }
}

class UMBaseTestCase : XCTestCase {}

class Evaluate: UMBaseTestCase
{
  func impasseTest()
  {
    let c1 = Creature(1, 2)
    let c2 = Creature(1, 2)
    let game = TemporaryCardDamageGame([c1, c2])
    XCTAssertEqual(-1, game.combat(0, 1))
    XCTAssertEqual(-1, game.combat(0, 1))
  }

  func temporaryMurderTest()
  {
    let c1 = Creature(1, 1)
    let c2 = Creature(2, 2)
    let game = TemporaryCardDamageGame([c1, c2])
    XCTAssertEqual(1, game.combat(0, 1))
  }

  func doubleMurderTest()
  {
    let c1 = Creature(2, 2)
    let c2 = Creature(2, 2)
    let game = TemporaryCardDamageGame([c1, c2])
    XCTAssertEqual(-1, game.combat(0, 1))
  }

  func permanentDamageDeathTest()
  {
    let c1 = Creature(1, 2)
    let c2 = Creature(1, 3)
    let game = PermanentCardDamage([c1, c2])
    XCTAssertEqual(-1, game.combat(0, 1))
    XCTAssertEqual(1, game.combat(0, 1))
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("impasseTest", impasseTest),
      ("temporaryMurderTest", temporaryMurderTest),
      ("doubleMurderTest", doubleMurderTest),
      ("permanentDamageDeathTest", permanentDamageDeathTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/4_BehavioralPatterns/10_Template/exercise.py
from unittest import TestCase
from abc import ABC


class Creature:
    def __init__(self, attack, health):
        self.health = health
        self.attack = attack


class CardGame(ABC):
    def __init__(self, creatures):
        self.creatures = creatures

    # return -1 if both creatures alive or both dead after combat
    # otherwise, return the _index_ of winning creature
    def combat(self, c1_index, c2_index):
        first = self.creatures[c1_index]
        second = self.creatures[c2_index]
        self.hit(first, second)
        self.hit(second, first)
        first_alive = first.health > 0
        second_alive = second.health > 0
        if first_alive == second_alive: return -1
        return c1_index if first_alive else c2_index

    def hit(self, attacker, defender):
        pass  # implement this in derived classes


class TemporaryDamageCardGame(CardGame):
    def hit(self, attacker, defender):
        old_health = defender.health
        defender.health -= attacker.attack
        if defender.health > 0:
            defender.health = old_health


class PermanentDamageCardGame(CardGame):
    def hit(self, attacker, defender):
        defender.health -= attacker.attack


class Evaluate(TestCase):
    def test_impasse(self):
        c1 = Creature(1, 2)
        c2 = Creature(1, 2)
        game = TemporaryDamageCardGame([c1, c2])
        self.assertEqual(-1, game.combat(0, 1), 'Combat should yield -1 since nobody died.')
        self.assertEqual(-1, game.combat(0, 1), 'Combat should yield -1 since nobody died.')

    def test_temporary_murder(self):
        c1 = Creature(1, 1)
        c2 = Creature(2, 2)
        game = TemporaryDamageCardGame([c1, c2])
        self.assertEqual(1, game.combat(0, 1))

    def test_double_murder(self):
        c1 = Creature(2, 1)
        c2 = Creature(2, 2)
        game = TemporaryDamageCardGame([c1, c2])
        self.assertEqual(-1, game.combat(0, 1))

    def test_permanent_damage_death(self):
        c1 = Creature(1, 2)
        c2 = Creature(1, 3)
        game = PermanentDamageCardGame([c1, c2])
        self.assertEqual(-1, game.combat(0, 1), 'Nobody should win this battle.')
        self.assertEqual(1, c1.health)
        self.assertEqual(2, c2.health)
        self.assertEqual(1, game.combat(0, 1), 'Creature at index 1 should win this')
